src/compiled_proof.rs

```diff
@@ -0,0 +1,442 @@
+            4u8 => Some(Ok(Command::HashPeak)),
+            5u8 => Some(Ok(Command::ToPeak)),
+            _ => Some(Err(Error::InvalidCommand(command))),
+        }
+    }
+}
+
+pub struct PackedLeaves<'a, T> {
+    index: usize,
+    data: &'a [u8],
+    t: PhantomData<T>,
+}
+
+impl<'a, T> PackedLeaves<'a, T> {
+    pub fn new(data: &'a [u8]) -> Self {
+        Self {
+            index: 0,
+            data,
+            t: PhantomData,
+        }
+    }
+}
+
+impl<'a, T: Packable> Iterator for PackedLeaves<'a, T> {
+    type Item = Result<(u64, T)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.data.len() {
+            return None;
+        }
+
+        if self.data.len() - self.index < 8 {
+            return Some(Err(Error::UnpackEof));
+        }
+
+        let pos = {
+            let mut buf = [0u8; 8];
+            buf.copy_from_slice(&self.data[self.index..self.index + 8]);
+            u64::from_le_bytes(buf)
+        };
+        self.index += 8;
+
+        match T::unpack(&self.data[self.index..]) {
+            Ok((item, size)) => {
+                self.index += size;
+                Some(Ok((pos, item)))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum StackItemType {
+    Node(u64, u8),
+    Peak,
+    Proof,
+}
+
+pub fn calculate_root<
+    T,
+    M: Merge<Item = T>,
+    I: Iterator<Item = Result<Command<T>>>,
+    IT: Iterator<Item = Result<(u64, T)>>,
+>(
+    commands: &mut I,
+    mmr_size: u64,
+    leaves: &mut IT,
+) -> Result<T> {
+    let mut last_leaf_pos = None;
+
+    let mut next_peak_info = PeakIterator::new(mmr_size);
+    // let mut leaf_index = 0;
+    let mut stack: Vec<(T, StackItemType)> = Vec::with_capacity(257);
+    for command in commands {
+        match command? {
+            Command::NextLeaf => {
+                let (pos, leaf_item) = leaves.next().ok_or(Error::CorruptedStack)??;
+                if pos_height_in_tree(pos) > 0 {
+                    return Err(Error::NodeProofsNotSupported);
+                }
+                if let Some(last_leaf_pos) = last_leaf_pos {
+                    // ensure leaves are sorted and unique
+                    if last_leaf_pos >= pos {
+                        return Err(Error::CorruptedProof);
+                    }
+                }
+                last_leaf_pos = Some(pos);
+                stack.push((leaf_item, StackItemType::Node(pos, 0)));
+            }
+            Command::Proof(proof) => {
+                stack.push((proof, StackItemType::Proof));
+            }
+            Command::Hash => {
+                if stack.len() < 2 {
+                    return Err(Error::CorruptedStack);
+                }
+                let (rhs, rhs_t) = stack.pop().unwrap();
+                let (lhs, lhs_t) = stack.pop().unwrap();
+                let (pos, height, next_pos, next_t, item, sibling_item) = match (lhs_t, &rhs_t) {
```

Comment:
Just an FYI: I have tried the path of introducing a `pos` function to `StackItemType` enum, so we don't need `next_pos` variable here. But surprisingly, a common `pos` function here actually will make performance worse(roughly 30000 more cycles on the overall 1583330 cycles), hence I'm keeping this `next_pos` variable.